addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationHandler.java

```diff
@@ -0,0 +1,591 @@
+                    sb.append(String.format("%s%n", id.getName()));
+                }
+            }
+            throw new PlugwiseInitializationException(
+                    String.format("Serial port '%s' could not be found. Available ports are:%n%s",
+                            configuration.getSerialPort(), sb));
+        }
+
+        try {
+            serialPort = portId.open("openHAB", 2000);
+            serialPort.addEventListener(this);
+            serialPort.notifyOnDataAvailable(true);
+            serialPort.setSerialPortParams(115200, SerialPort.DATABITS_8, SerialPort.STOPBITS_1,
+                    SerialPort.PARITY_NONE);
+            outputChannel = Channels.newChannel(serialPort.getOutputStream());
+        } catch (IOException e) {
+            throw new PlugwiseInitializationException("Failed to get serial port output stream", e);
+        } catch (PortInUseException e) {
+            throw new PlugwiseInitializationException("Serial port already in use", e);
+        } catch (TooManyListenersException e) {
+            throw new PlugwiseInitializationException("Failed to add serial port listener", e);
+        } catch (UnsupportedCommOperationException e) {
+            throw new PlugwiseInitializationException("Failed to set serial port parameters", e);
+        }
+    }
+
+    public boolean isExistingMessageListener(PlugwiseMessageListener listener) {
+        for (FilteredMessageListener filteredListener : filteredListeners) {
+            if (filteredListener.listener.equals(listener)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean isExistingMessageListener(PlugwiseMessageListener listener, MACAddress macAddress) {
+        for (FilteredMessageListener filteredListener : filteredListeners) {
+            if (filteredListener.listener.equals(listener) && macAddress.equals(filteredListener.macAddress)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void notifyListeners(Message message) {
+        for (FilteredMessageListener filteredListener : filteredListeners) {
+            if (filteredListener.matches(message)) {
+                try {
+                    filteredListener.listener.handleReponseMessage(message);
+                } catch (Exception e) {
+                    logger.warn("Listener failed to handle message: {}", message, e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Parse a buffer into a Message and put it in the appropriate queue for further processing
+     *
+     * @param readBuffer - the string to parse
+     */
+    private void parseAndQueue(ByteBuffer readBuffer) {
+        if (readBuffer != null) {
+
+            String response = new String(readBuffer.array(), 0, readBuffer.limit());
+            response = StringUtils.chomp(response);
+
+            Matcher matcher = RESPONSE_PATTERN.matcher(response);
```

Comment:
The number of messages depends on the network size and what kind channel configuration is used. I haven't run into any performance issues with it on my RPi3. It also helps that the pattern is created only once and reused for matching.

This code now lives in the PlugwiseMessageProcessor.